exonum-java-binding/core/rust/src/proxy/node.rs

```diff
@@ -69,34 +63,28 @@ impl NodeContext {
 
     #[doc(hidden)]
     pub fn create_snapshot(&self) -> Box<dyn Snapshot> {
-        self.blockchain.snapshot()
+        self.api_context.snapshot()
     }
 
     #[doc(hidden)]
     pub fn public_key(&self) -> PublicKey {
-        self.public_key
+        self.api_context.service_keypair().0.clone()
     }
 
     #[doc(hidden)]
-    pub fn submit(&self, transaction: RawTransaction) -> Result<Hash, failure::Error> {
-        let service_id = transaction.service_id();
-
-        if !self.blockchain.service_map().contains_key(&service_id) {
-            return Err(format_err!(
-                "Unable to broadcast transaction: service(ID={}) not found",
-                service_id
-            ));
-        }
-
-        let msg = Message::sign_transaction(
-            transaction.service_transaction(),
-            service_id,
-            self.blockchain.service_keypair.0,
-            &self.blockchain.service_keypair.1,
+    pub fn submit(&self, tx: AnyTx) -> Result<Hash, failure::Error> {
+        let (pub_key, secret_key) = self.api_context.service_keypair();
+        let signed = SignedMessage::new(
+            tx.to_bytes(),
+            pub_key.to_owned(),
+            secret_key,
         );
-        let tx_hash = msg.hash();
 
-        self.transaction_sender.broadcast_transaction(msg)?;
+        let tx_hash = signed.object_hash();
+        // TODO(ECR-3679): check Core behaviour/any errors on service inactivity
+        self.api_context.sender().broadcast_transaction(
+            signed.into_verified()?
```

Comment:
I don't think we shall *verify* a transaction message that has been just created.
            Can we use from_value instead?